File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
class Solution:
    def mergeKLists(self, lists: list[Optional[ListNode]]) -> Optional[ListNode]:
        nodes = []
        for lst in lists:
            while lst:
                nodes.append(lst)
                lst = lst.next
        nodes.sort(key=lambda x: x.val)
        head = ListNode(0)  # Dummy head
        current = head
        for node in nodes:
            current.next = node
            current = current.next
        current.next = None  
        return head.next

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a.val));
        for (ListNode node : lists) {
            if (node != null) {
                minHeap.offer(node);
            }
        }
        
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        while (!minHeap.isEmpty()) {
            ListNode smallest = minHeap.poll(); 
            current.next = smallest; 
            current = current.next; 
            if (smallest.next != null) {
                minHeap.offer(smallest.next);
            }
        }
        
        return dummy.next; 
    }
}
